File: src/web/stream.js
const ffmpeg = require('fluent-ffmpeg');

const channels = require('../../channels.json');

const stream = (req, res) => {
  const { channelNum } = req.params;
  const { transcode } = req.query;
  const channel = channels[channelNum - 1];

  console.log(`[STREAM] Starting '${channel.name}'`);

  res.writeHead(200, {
    'Content-Type': 'video/mp2t',
  });

  let ffmpegStream = ffmpeg(channel.url);

  // Proxy
  if (channel.proxy) {
    const { host, username, password } = channel.proxy;
    ffmpegStream = ffmpegStream
      .addInputOption(`-http_proxy http://${username}:${password}@${host}`);
  }

  // Video acceleration
  if (process.platform === 'darwin') {
    ffmpegStream = ffmpegStream
      // .addInputOption('-hwaccel videotoolbox')
      .videoCodec('h264_videotoolbox')
      .addOutputOption('-b:v 10M');
  } else if (process.platform === 'linux') {
    ffmpegStream = ffmpegStream
      .addInputOption('-hwaccel vaapi')
      .videoCodec('h264_vaapi');
  } else {
    ffmpegStream = ffmpegStream
      .videoCodec('libx264')
      .preset('superfast');
  }

  ffmpegStream = ffmpegStream
    .audioCodec('aac')
    .addOutputOption('-b:a 256k')
    .addOutputOption('-copyinkf')
    .addOutputOption('-metadata service_provider=AMAZING')
    .addOutputOption(`-metadata service_name=${channel.name.replace(/\s/g, '-')}`)
    .addOutputOption('-tune zerolatency')
    .addOutputOption('-mbd rd')
    .addOutputOption('-flags +ilme+ildct')
    .addOutputOption('-fflags +genpts')
    .outputFormat('mpegts');

  switch (transcode) {
    case 'mobile':
    case 'internet720':
      ffmpegStream = ffmpegStream
        .size('1280x720')
        .outputFPS(30);
      break;
    // case 'internet480':
    //   break;
    // case 'internet360':
    //   break;
    // case 'internet240':
    //   break;
    default:
      break;
  }

  ffmpegStream.on('error', console.error);

  ffmpegStream.pipe(res);
};

module.exports = {
  stream,
};

File: routes/stream.go
package routes

import (
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strconv"
	"strings"

	"github.com/duncanleo/plex-dvr-hls/config"
	"github.com/gin-gonic/gin"
)

func Stream(c *gin.Context) {
	var channelIDStr = c.Param("channelID")
	channelID, err := strconv.Atoi(channelIDStr)
	if err != nil {
		c.Status(http.StatusBadRequest)
		return
	}

	var channel = config.Channels[channelID-1]
	var transcode = c.Query("transcode")

	log.Printf("[STREAM] Starting '%s'\n", channel.Name)

	c.Header("Content-Type", "video/mp2t")

	var ffmpegArgs []string

	if channel.ProxyConfig != nil {
		ffmpegArgs = append(
			ffmpegArgs,
			"-http_proxy",
			fmt.Sprintf("http://%s:%s@%s", channel.ProxyConfig.Username, channel.ProxyConfig.Password, channel.ProxyConfig.Host),
		)
	}

	switch config.Cfg.GetEncoderProfile() {
	case config.EncoderProfileVAAPI:
		ffmpegArgs = append(
			ffmpegArgs,
			"-vaapi_device",
			"/dev/dri/renderD128",
			"-hwaccel",
			"vaapi",
			"-hwaccel_output_format",
			"vaapi",
		)
	case config.EncoderProfileVideoToolbox:
		ffmpegArgs = append(
			ffmpegArgs,
			"-hwaccel",
			"videotoolbox",
		)
	}

	ffmpegArgs = append(
		ffmpegArgs,
		"-i",
		channel.URL,
	)

	if channel.DisableTranscode {
		ffmpegArgs = append(
			ffmpegArgs,
			"-c:v",
			"copy",
		)
	} else {
		switch config.Cfg.GetEncoderProfile() {
		case config.EncoderProfileVideoToolbox:
			ffmpegArgs = append(
				ffmpegArgs,
				"-c:v",
				"h264_videotoolbox",
			)
			break
		case config.EncoderProfileVAAPI:
			ffmpegArgs = append(
				ffmpegArgs,
				"-c:v",
				"h264_vaapi",
				"-vf",
				"format=nv12|vaapi,hwupload",
			)
			break
		case config.EncoderProfileOMX:
			ffmpegArgs = append(
				ffmpegArgs,
				"-c:v",
				"h264_omx",
			)
			break
		default:
			ffmpegArgs = append(
				ffmpegArgs,
				"-c:v",
				"libx264",
				"-preset",
				"superfast",
			)
			break
		}
	}

	ffmpegArgs = append(
		ffmpegArgs,
		"-b:a",
		"256k",
		"-copyinkf",
		"-metadata",
		"service_provider=AMAZING",
		"-metadata",
		fmt.Sprintf("service_name=%s", strings.ReplaceAll(channel.Name, " ", "")),
		"-tune",
		"zerolatency",
		"-mbd",
		"rd",
		"-flags",
		"+ilme+ildct",
		"-fflags",
		"+genpts",
	)

	switch transcode {
	case "mobile":
	case "internet720":
		ffmpegArgs = append(
			ffmpegArgs,
			"-s",
			"1280x720",
			"-r",
			"30",
		)
		break
	}

	ffmpegArgs = append(
		ffmpegArgs,
		"-f",
		"mpegts",
		"pipe:1",
	)

	var ffmpegProcess = exec.Command(
		"ffmpeg",
		ffmpegArgs...,
	)

	outPipe, err := ffmpegProcess.StdoutPipe()
	if err != nil {
		log.Println(err)
		c.Status(http.StatusInternalServerError)
		return
	}

	ffmpegProcess.Stderr = os.Stdout

	err = ffmpegProcess.Start()
	if err != nil {
		log.Println(err)
		c.Status(http.StatusInternalServerError)
		return
	}

	c.Stream(func(w io.Writer) bool {
		_, err := io.Copy(w, outPipe)

		if err != nil {
			log.Println(err)
			return true
		}

		if ffmpegProcess.Process != nil {
			ffmpegProcess.Process.Kill()
		}
		return true
	})

}
